### PR TITLE
Run and test benchmarks against C# as well

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

Turns out, there are some APIs that are not covered by any of our tests, except benchmarks.

Benchmarks are executed as both actual benchmarks and as integration tests for Rust, but not for C#. This PR fixes that to ensure that both Rust and C# are tested and measured on the CI from now on.

It also makes some adjustments to integration tests themselves, in particular fixes an issue where in integration tests reducers would be called but the response would never be read, so if a reducer panics, the tests would still succeed. Additionally, on failures the SpacetimeDB logs will be read and added to the error message so that you get a useful stacktrace from the module itself.

I also had to adjust some benchmark numbers (in particular, for `ia_loop`) since they were already very slow on CI for Rust, but become even more so for C#. The new loads are much lower, but should still be very usable for benchmark comparisons.

The only remaining issue is that the benchmark names have changed, and the benchmark comparison tool tries to compare mismatched pairs - I'm not sure what the solution to this is. cc @kazimuth 

# API and ABI breaking changes

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

**Comment by RReverser** [comment]
benchmarks please

**Comment by RReverser** [comment]
benchmarks please

File: crates/testing/src/modules.rs [hunk 1]
```diff
@@ -223,3 +226,43 @@ pub struct LoggerRecord {
     pub line_number: Option<u32>,
     pub message: String,
 }
+
+const COMPILATION_MODE: CompilationMode = if cfg!(debug_assertions) {
```
Comment (kazimuth) [inline crates/testing/src/modules.rs:230]: Do we know if the `debug_assertions` feature is currently enabled in the transitive dependencies of the `bench` crate?

File: crates/core/src/host/module_host.rs [hunk 2]
```diff
@@ -238,14 +238,12 @@ impl ModuleInfo {
...
-/// Invariant: the reducer names are in alphabetical order.
+/// Invariant: the reducer names are in the same order as they were declared in the `ModuleDef`.
 pub struct ReducersMap(IndexSet<Box<str>>);
 
 impl<'a> FromIterator<&'a str> for ReducersMap {
     fn from_iter<T: IntoIterator<Item = &'a str>>(iter: T) -> Self {
-        let mut sorted = Vec::from_iter(iter);
-        sorted.sort();
-        ReducersMap(sorted.into_iter().map_into().collect())
+        Self(iter.into_iter().map_into().collect())
```
Comment (kazimuth) [inline crates/core/src/host/module_host.rs:246]: If this PR is delayed we should do a separate PR with this change and merge it right away cause it's a straight up correctness fix.

File: crates/testing/src/modules.rs (hunk 1 above)
Comment (RReverser) [reply crates/testing/src/modules.rs:230]: No, but it doesn't have to be - the point here is to use that mode only in integration test mode for easier debugging (it helped me with a stacktrace).

**Comment by RReverser** [comment]
@kazimuth FWIW this is marked as draft because it's not quite ready to look at. I only created a PR because without it our CI doesn't run on branches :)

**Comment by kazimuth** [comment]
Well, looks good to me so far :)

**Comment by RReverser** [comment]
benchmarks please

File: crates/core/src/host/module_host.rs (hunk 2 above)
Comment (RReverser) [reply crates/core/src/host/module_host.rs]: Done in #1980.

**Comment by RReverser** [comment]
benchmarks please

**Comment by RReverser** [comment]
The failure here is the one that will be fixed by #1987.

**Comment by RReverser** [comment]
Copied relevant change myself to unblock the CI.

File: crates/bench/src/spacetime_module.rs [hunk 3]
```diff
@@ -42,26 +25,27 @@ lazy_static::lazy_static! {
 ///
 /// See the doc comment there for information on the formatting expected for
 /// table and reducer names.
-pub struct SpacetimeModule {
+pub struct SpacetimeModule<L> {
```
Comment (kazimuth) [inline crates/bench/src/spacetime_module.rs:28]: neat, I like this.

**Review by gefjon — APPROVED** [review]

**Comment by gefjon** [comment]
Internal tests failures appear to be spurious/unrelated, but let's wait to merge until they're fixed just in case.
